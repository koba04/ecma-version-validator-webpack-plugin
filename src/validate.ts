import { parse, defaultOptions } from "acorn";
import { format } from "./format";

export type ECMAVersion = typeof defaultOptions.ecmaVersion;

export type ErrorMap = Map<
  string,
  {
    message: string;
    source: string;
    position: {
      line: number;
      column: number;
    } | null;
  }
>;

export const validate = (
  assets: { [file: string]: { _value: string } },
  options: { ecmaVersion: ECMAVersion }
) => {
  const { ecmaVersion } = options;
  const errors: ErrorMap = new Map();
  Object.entries(assets).forEach(([file, source]) => {
    try {
      parse(source._value, { ecmaVersion });
    } catch (e) {
      const message = e.toString();
      const match = message.match(/.*?\((\d+):(\d+)\)$/);
      const position = match ? { line: +match[1], column: +match[2] } : null;
      errors.set(file, { message, source: source._value, position });
    }
  });
  if (errors.size > 0) {
    throw new Error(format(errors));
  }
};
