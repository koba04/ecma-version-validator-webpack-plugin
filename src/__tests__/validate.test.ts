import { validate } from "../validate";

describe("validate", () => {
  describe("success", () => {
    it("should not thrown any error if input sources are valid", () => {
      expect(() => {
        validate(
          {
            "foo.js": { _value: "console.log(123);" },
            "bar.js": { _value: "var bar = 123" },
          },
          { ecmaVersion: 5 }
        );
      }).not.toThrow();
    });
    it("should not throw any error if input sources are valid with a passed ecma version", () => {
      expect(() => {
        validate(
          {
            "foo.js": { _value: "const foo = () => console.log(123);" },
            "bar.js": { _value: "class Bar {}" },
          },
          { ecmaVersion: 2015 }
        );
      }).not.toThrow();
    });
  });
  describe("failed", () => {
    it("should throw an error includes all files having any error", () => {
      expect(() => {
        validate(
          {
            "foo.js": {
              _value: "var foo = 1;\nvar foo = () => console.log(123);",
            },
            "bar.js": { _value: "var bar = 1;\nclass Bar {}" },
            "baz.js": { _value: "var baz = true" },
          },
          { ecmaVersion: 5 }
        );
      }).toThrowErrorMatchingSnapshot();
    });
  });
});
