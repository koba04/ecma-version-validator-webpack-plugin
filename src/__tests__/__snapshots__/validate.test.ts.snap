// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate failed should throw an error includes all files having any error 1`] = `
"[foo.js]:SyntaxError: Unexpected token (2:11)
[2:11] var foo = () => console.log(123);
[bar.js]:SyntaxError: The keyword 'class' is reserved (2:0)
[2:0] class Bar {}
"
`;
